File: 207-course-schedule/207. Course Schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool: 
        indegree = [0] * numCourses
        dic = defaultdict(list)
        for u, v in prerequisites:
            indegree[u] += 1
            dic[v].append(u)
        
        visited = set()
        queue = deque([i for i in range(numCourses) if indegree[i] == 0])
        # print(queue)
        while queue:
            node = queue.popleft()
            if indegree[node] == 0:
                visited.add(node)
                for nei in dic[node]:
                    indegree[nei] -= 1
                    if indegree[nei] == 0:
                        queue.append(nei)
        return len(visited) == numCourses

File: 207-course-schedule/207. Course Schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, int> indegree;
        unordered_map<int, vector<int>> graph;

        for (auto& arr : prerequisites) {
            int course = arr[0], preq = arr[1];
            graph[preq].push_back(course);
            indegree[course]++;
        }
        cout << "reached" << endl;
        queue<int> q;
        unordered_set<int> taken_courses;
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                q.push(i);
                taken_courses.insert(i);
            }
        }

        while (!q.empty()) {
            int course = q.front();
            q.pop();
            for (auto& nei : graph[course]) {
                indegree[nei]--;
                if (indegree[nei] == 0) {
                    q.push(nei);
                    taken_courses.insert(nei);
                }
            }
        }

        return taken_courses.size() == numCourses;
    }
};